### ISSUE TITLE
More a Question than an issue per say.

### ISSUE THREAD

**Issue by PBadchild** [snobu/destreamer#124]
Hi, SORRY if this is the wrong place for this I had no idea where else to ask or to go.

I was wondering if I use the " --vcodec libx265 " 

Is the download supposed to be faster or slower or the same ? I though it would be faster cause thats one of the advantages of the codec (from what I understand). But I tried it and it seems to be longer. I'm not sure why but my guess is its longer to copy video than to recode it with new codec than to just copy and encode it with the same ?

I know this might be a stupid question and I might not even make sense haha but please give my a break I have been working to make this work for me with absolutely no prior knowledge of programmig or coding.

But as there was no executable version of destreamer and I really needed it so I decided to get in way over my head hahaha

BUT HEY it worked ! I followed ## Prereqs and then the ## How to build. That was actually the easy part haha. Then I manneged to make it work once for one individual video but it was in GitBash 🤦‍♂️ So I had no progression bar. So I tried to make it work in Powershell and after a few a hours of research got it to work !  

In any case,  @snobu  Thanks a lot for this little gem !

**Comment by molikuner** [comment]
Hey @PBadchild,

TL;DR
Specifying a encoder other than `copy` forces ffmpeg to re-encode the data, but you get a output with a guaranteed codec which isn't the case with `copy` since that doesn't change the data at all.

---

> I'm not sure why but my guess is its longer to copy video than to recode it with new codec than to just copy and encode it with the same ?

Almost 🙂. When you specify a encoder (other than `copy`), ffmpeg handles the downloaded data with the following steps:

```
 _______              ______________
|       |            |              |
| input |  demuxer   | encoded data |   decoder
| file  | ---------> | packets      | -----+
|_______|            |______________|      |
                                           v
                                       _________
                                      |         |
                                      | decoded |
                                      | frames  |
                                      |_________|
 ________             ______________       |
|        |           |              |      |
| output | <-------- | encoded data | <----+
| file   |   muxer   | packets      |   encoder
|________|           |______________|

[1]
```
As you can see, that process includes decoding and encoding the video. By default `destreamer` specifies `copy` as encoder, so the process gets simplified to:
```
 _______              ______________            ________
|       |            |              |          |        |
| input |  demuxer   | encoded data |  muxer   | output |
| file  | ---------> | packets      | -------> | file   |
|_______|            |______________|          |________|

[2]
```
As you can see, the decode and encode isn't even the happening. That's why it's so much faster. So even if you use a fast encodable codec, this is probably slower than the default `copy`. But `copy` doesn't allow any safe assumption about the output, since it literally just copies the input into the new container. So if the input codec is something weird, the output will have the same codec. So now you need to decide between performance and extreme safe output. The reason why, at least in my opinion, `copy` is okay in this case is, that MS Streams will probably not use some weird codec because the web browser needs to be able to play it. And if possible, the web browser could also use hardware acceleration for that, which is probably only available for common codecs. So it's probably in the best interest of Microsoft to use a good standard for that.

To be fair, you can probably achieve better speeds with hardware encoding than software encoding. But the support for that is limited and very hardware depended. See this [link](https://trac.ffmpeg.org/wiki/HWAccelIntro#NVENC) for more information about that.

---
[1] Taken from https://ffmpeg.org/ffmpeg.html#Detailed-description
[2] Taken from https://ffmpeg.org/ffmpeg.html#Stream-copy

**Comment by snobu** [comment]
@molikuner thank you for the beautiful ascii art and writeup.
@PBadchild, this is now implemented via the `--format` command line argument. See the README for more.

**Comment by PBadchild** [comment]
Wow @molikuner Thanks you so much ! Your diagram is so clear I got it all ! Sorry for the late answer I did not think someone would take the time to answer ! Thanks again ! :) You guys do great work I'm very gratefull !
